### PR TITLE
[IMP] mass_mailing, *: improve mailing stat buttons

### PR DESCRIPTION
**Pull request by pvhelden**
*mass_mailing_crm, mass_mailing_sale 

Currently, the Mailing stat buttons have a few problems.
The business reporting smart buttons send the user to a list of linked records,
which doesn't provide much value (e.g. what is the user supposed to do with a
list of 12345 leads?). They also show wrong values depending on access rights.
The Mailing KPI smart buttons lead to a blank screen if there's no data to show.

In this commit, we add/adapt all action helpers to help users understand what is
going on. For the business reporting smart buttons we instead send users to
relevant reporting actions, so they can make sense of the data we showed before.
The values of the business reporting smart buttons are now calculated as a super
user.
We also change some domains from the whole UTM Sale Domain to only using
Source ID, which simplifies computation while staying accurate.

task-2604515

### CODE REVIEW

File: addons/mass_mailing/models/mailing.py [hunk 1]
```diff
@@ -518,12 +518,17 @@ def _action_view_traces_filtered(self, view_filter):
 
     def action_view_clicked(self):
         model_name = self.env['ir.model']._get('link.tracker').display_name
+        helper = (
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:521]: Just for clarity, I would define helper_header and helper_message, and use them below :) code would be easier to follow that way

File: addons/mass_mailing/models/mailing.py [hunk 2]
```diff
@@ -518,12 +518,17 @@ def _action_view_traces_filtered(self, view_filter):
...
+            _("No %s clicked your mailing yet!", model_name),
+            _("Link Trackers will measure how many times each link is clicked as well as the proportion of TargetModel who clicked at least once in your mailing.")
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:523]: Linting: maybe multi-line this sentence (going outside of github review = too long line ^^)

File: addons/mass_mailing/models/mailing.py [hunk 3]
```diff
@@ -518,12 +518,17 @@ def _action_view_traces_filtered(self, view_filter):
...
+        )
         return {
             'name': model_name,
             'type': 'ir.actions.act_window',
             'view_mode': 'tree',
             'res_model': 'link.tracker',
             'domain': [('mass_mailing_id.id', '=', self.id)],
+            'help': '<p class="o_view_nocontent_smiling_face">' + helper[0] + '</p><p>' + helper[1] + '</p>',
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:531]: We could probably use string formatting to make it nicer (aka something like ``<p>%(helper_header)s ...``) .

File: addons/mass_mailing/models/mailing.py [hunk 4]
```diff
@@ -540,28 +545,48 @@ def action_view_delivered(self):
...
+        action = {
             'name': model_name,
             'type': 'ir.actions.act_window',
             'view_mode': 'tree',
             'res_model': self.mailing_model_real,
             'domain': [('id', 'in', res_ids)],
             'context': dict(self._context, create=False)
         }
+        if helper:
+            action['help'] = '<p class="o_view_nocontent_smiling_face">' + helper[0] + '</p><p>' + helper[1] + '</p>'
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:588]: Same global remark: would define helper_header / helper_message and use string formatting here, just for code clarity

File: addons/mass_mailing_crm/models/mailing_mailing.py [hunk 5]
```diff
@@ -24,12 +24,17 @@ def _compute_crm_lead_count(self):
             mass_mailing.crm_lead_count = mapped_data.get(mass_mailing.source_id.id, 0)
 
     def action_redirect_to_leads_and_opportunities(self):
-        view = 'crm.crm_lead_all_leads' if self.use_leads else 'crm.crm_lead_opportunities'
-        action = self.env.ref(view).sudo().read()[0]
-        action['view_mode'] = 'tree,kanban,graph,pivot,form,calendar'
-        action['domain'] = [('source_id', 'in', self.source_id.ids)]
-        action['context'] = {'active_test': False, 'create': False}
-        return action
+        text = "Leads" if self.use_leads else "Opportunities"
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_crm/models/mailing_mailing.py:27]: Don't forget translation marks

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml [hunk 6]
```diff
@@ -22,4 +22,17 @@
     		</xpath>
         </field>
     </record>
+
+    <record id="mailing_mailing_lead_view_search" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_crm/views/mailing_mailing_views.xml:26]: What is the purpose of this new view, primary, with priority 1 (aka will be used by default) ? Just see a by day instead of a by month ? Is that really relevant ? If we really want a specific view for that, maybe set a higher priority (aka less default) and ensure action in mass mailing uses it (and not other actions).

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 7]
```diff
@@ -15,47 +15,46 @@ class MassMailing(models.Model):
 
     @api.depends('mailing_domain')
     def _compute_sale_quotation_count(self):
-        has_so_access = self.env['sale.order'].check_access_rights('read', raise_exception=False)
-        if not has_so_access:
-            self.sale_quotation_count = 0
-            return
         for mass_mailing in self:
-            mass_mailing.sale_quotation_count = self.env['sale.order'].search_count(mass_mailing._get_sale_utm_domain())
+            mass_mailing.sale_quotation_count = self.env['sale.order'].sudo().search_count(mass_mailing._get_sale_utm_domain())
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:19]: I am wondering if we should not simply search on source, instead of the whole domain... maybe to check with Lna and Dbe to be sure we are consistent, but it could simplify a lot the computation while still being accurate.

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 8]
```diff
@@ -15,47 +15,46 @@ class MassMailing(models.Model):
...
 
     def action_redirect_to_quotations(self):
-        action = self.env["ir.actions.actions"]._for_xml_id("sale.action_quotations_with_onboarding")
-        action['domain'] = self._get_sale_utm_domain()
-        action['context'] = {'create': False}
-        return action
+        return {
+            'context': {'create': False, 'search_default_group_by_date': True},
+            'domain': self._get_sale_utm_domain(),
+            'help': '<p class="o_view_nocontent_smiling_face">' + _("No Quotations yet!") + '</p><p>' +
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:35]: Maybe avoid hardcoded + and use string formatting, better syntax :)

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 9]
```diff
@@ -15,47 +15,46 @@ class MassMailing(models.Model):
...
-        if not self.user_has_groups('sales_team.group_sale_salesman') or not self.user_has_groups('account.group_account_invoice'):
-            self.sale_invoiced_amount = 0
-            return
         for mass_mailing in self:
             domain = expression.AND([
                 mass_mailing._get_sale_utm_domain(),
                 [('state', 'not in', ['draft', 'cancel'])]
             ])
-            moves = self.env['account.move'].search_read(domain, ['amount_untaxed_signed'])
+            moves = self.env['account.move'].sudo().search_read(domain, ['amount_untaxed_signed'])
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:28]: Same here, proposal to use only the source and get rid of sale utm domain usage (complex for few real use cases I think)

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 10]
```diff
@@ -15,47 +15,46 @@ class MassMailing(models.Model):
...
-            'edit': False,
-            'view_no_maturity': True
+        return {
+            'context': {'create': False, 'search_default_group_by_invoice_date': True},
+            'domain': [
+                ('id', 'in', self.env['account.move'].search(self._get_sale_utm_domain()).ids),
+                ('move_type', 'in', ('out_invoice', 'out_refund', 'in_invoice', 'in_refund', 'out_receipt', 'in_receipt')),
+                ('state', 'not in', ['draft', 'cancel'])
+            ],
+            'help': '<p class="o_view_nocontent_smiling_face">' + _("No Revenues yet!") + '</p><p>' +
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:51]: Same, would use string formatting

File: addons/mass_mailing_sale/views/mailing_mailing_views.xml [hunk 11]
```diff
@@ -25,4 +25,29 @@
     		</xpath>
         </field>
     </record>
+
+    <record id="mailing_mailing_invoice_view_search" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/views/mailing_mailing_views.xml:29]: Same remark, you are adding a new default view. Would propagate lead view comments on this one :)

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml [hunk 12]
```diff
@@ -22,4 +22,17 @@
...
+        <field name="name">mailing.mailing.lead.report.search</field>
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_crm/views/mailing_mailing_views.xml:27]: Guidelines for naming are about used model, not "source model". For example here it could be ``crm.lead.view.search.from.mailing`` (and xml id would be the same with underscores instead of dots).

**Review by tde-banana-odoo — COMMENTED** [review]
Validation ! Thanks for your PR !

Some remarks (translations, string formatting), some questions :) .

Cheers !

File: addons/mass_mailing_crm/models/mailing_mailing.py [hunk 13]
```diff
@@ -24,12 +24,20 @@ def _compute_crm_lead_count(self):
...
-        action['context'] = {'active_test': False, 'create': False}
-        return action
+        text = _("Leads") if self.use_leads else _("Opportunities")
+        helper_header = _(f"No {text} yet!")
+        helper_message = _("Make sure to include a link to a form or a reply-to an " +
+                           f"alias to generate {text} from this mailing.")
+        return {
+            'context': {'active_test': False, 'create': False, 'search_default_day': True},
+            'domain': [('source_id', 'in', self.source_id.ids)],
+            'help': f'<p class="o_view_nocontent_smiling_face">{helper_header}</p><p>{helper_message}</p>',
```
Comment (dbeguin) [inline addons/mass_mailing_crm/models/mailing_mailing.py:34]: fstring doesn't work with translation. please use `_("your %s is your brol", brol)` instead

also, not sure this sentence is super clear : 
`Make sure to include a link to a form or a reply-to an alias to generate {text} from this mailing.`
Maybe reword a little ? @LNAodoo

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml [hunk 14]
```diff
@@ -22,4 +22,15 @@
...
+        <field name="name">crm.lead.view.search.from.mailing</field>
+        <field name="model">crm.lead</field>
+        <field name="inherit_id" ref="crm.view_crm_case_opportunities_filter"/>
+        <field name="arch" type="xml">
+            <xpath expr="//group/filter[@name='month']" position="replace">
+                <filter string="Creation Date" context="{'group_by':'create_date:day'}" name="day"/>
+            </xpath>
+        </field>
+    </record>
 </odoo>
```
Comment (dbeguin) [inline addons/mass_mailing_crm/views/mailing_mailing_views.xml:36]: while we are at it, add the last empty line to make the DBElinter happy :smile:

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml [hunk 15]
```diff
@@ -22,4 +22,15 @@
     		</xpath>
         </field>
     </record>
+
+    <record id="crm_lead_view_search_from_mailing" model="ir.ui.view">
+        <field name="name">crm.lead.view.search.from.mailing</field>
```
Comment (dbeguin) [inline addons/mass_mailing_crm/views/mailing_mailing_views.xml:27]: guidelines : 
```
<record id="crm_lead_view_search" model="ir.ui.view">
    <field name="name">crm.lead.view.search.inherit.mass.mailing</field>
```

File: addons/mass_mailing_sale/views/mailing_mailing_views.xml [hunk 16]
```diff
@@ -25,4 +25,29 @@
...
+        <field name="model">sale.report</field>
+        <field name="mode">primary</field>
+        <field name="inherit_id" ref="sale.view_order_product_search"/>
+        <field name="arch" type="xml">
+            <xpath expr="//group/filter[@name='date']" position="replace">
+                <filter string="Order Date" name="group_by_date" context="{'group_by':'date:day'}"/>
+            </xpath>
+        </field>
+    </record>
 </odoo>
```
Comment (dbeguin) [inline addons/mass_mailing_sale/views/mailing_mailing_views.xml:53]: same here for DBElinting

File: addons/mass_mailing_sale/views/mailing_mailing_views.xml [hunk 17]
```diff
@@ -25,4 +25,29 @@
     		</xpath>
         </field>
     </record>
+
+    <record id="mailing_mailing_invoice_view_search" model="ir.ui.view">
+        <field name="name">mailing.mailing.invoice.report.search</field>
+        <field name="model">account.invoice.report</field>
```
Comment (dbeguin) [inline addons/mass_mailing_sale/views/mailing_mailing_views.xml:31]: guidelines : 
```
<record id="module2.model_view_form" model="ir.ui.view">
    <field name="name">model.view.form.module2</field>
```

File: addons/mass_mailing_sale/views/mailing_mailing_views.xml [hunk 18]
```diff
@@ -25,4 +25,29 @@
...
+        <field name="arch" type="xml">
+            <xpath expr="//group/filter[@name='invoice_date']" position="replace">
+                <filter string="Date" name="group_by_invoice_date" context="{'group_by':'invoice_date:week'}"/>
+            </xpath>
+        </field>
+    </record>
+
+    <record id="mailing_mailing_quotation_view_search" model="ir.ui.view">
+        <field name="name">mailing.mailing.quotation.report.search</field>
+        <field name="model">sale.report</field>
```
Comment (dbeguin) [inline addons/mass_mailing_sale/views/mailing_mailing_views.xml:44]: same here

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 19]
```diff
@@ -15,47 +15,47 @@ class MassMailing(models.Model):
...
     def action_redirect_to_quotations(self):
-        action = self.env["ir.actions.actions"]._for_xml_id("sale.action_quotations_with_onboarding")
-        action['domain'] = self._get_sale_utm_domain()
-        action['context'] = {'create': False}
-        return action
+        return {
+            'context': {'create': False, 'search_default_group_by_date': True},
+            'domain': self._get_sale_utm_domain(),
+            'help': '<p class="o_view_nocontent_smiling_face">' + _("No Quotations yet!") + '</p><p>' +
+                    _("Quotations will appear here once your customers add products to their Carts or when your sales reps assign this mailing.") + '</p>',
```
Comment (dbeguin) [inline addons/mass_mailing_sale/models/mailing_mailing.py:36]: we prefer to use %s instead of string concatenation.

**Review by dbeguin — COMMENTED** [review]
Ola ! some additional comments

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml (hunk 15 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing_crm/views/mailing_mailing_views.xml:27]: +1, indeed my previous comment was either a bit Falsy, either the view was primary. If it is not primary, it just improve the existing one, so no need to add a suffix.

Reason:
base view should be called ``crm.crm_lead_view_tree`` (current naming is historic)
inherit view will be called ``mass_mailing_crm.crm_lead_view_tree``
-> we know those are making a single global view to display, automatic prefix in xml ids tells the module in which a specific view is defined

If view is primary, this is a new view. It has to have a name telling its usage. Aka ``crm_lead_view_search_from_mailing`` for example if it is used only when coming from a mailing.

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml [hunk 20]
```diff
@@ -22,4 +22,15 @@
...
+        <field name="model">crm.lead</field>
+        <field name="inherit_id" ref="crm.view_crm_case_opportunities_filter"/>
+        <field name="arch" type="xml">
+            <xpath expr="//group/filter[@name='month']" position="replace">
+                <filter string="Creation Date" context="{'group_by':'create_date:day'}" name="day"/>
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_crm/views/mailing_mailing_views.xml:32]: Still wondering why we change this filter globally with mass mailing. Either this is primary (specific to mass mailing), either this should be done in base CRM module. Strange that having the bridge of mass mailing and crm changes the way "creation date" filter works.

Comment (SwagSamaSempai) [reply addons/mass_mailing_crm/views/mailing_mailing_views.xml:32]: This should absolutely be primary, it is indeed specific to Mass Mailing, and shouldn't affect CRM itself.

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml [hunk 21]
```diff
@@ -22,4 +21,17 @@
...
         </field>
     </record>
-</odoo>
\ No newline at end of file
+
+    <record id="crm_lead_view_search" model="ir.ui.view">
+        <field name="name">crm.lead.view.search.inherit.mass.mailing</field>
+        <field name="model">crm.lead</field>
+        <field name="mode">primary</field>
+        <field name="priority">9001</field>
```
Comment (awa-odoo) [inline addons/mass_mailing_crm/views/mailing_mailing_views.xml:29]: That... is a lot :smile:

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 22]
```diff
@@ -10,52 +10,59 @@ class MassMailing(models.Model):
...
+    sale_invoiced_amount = fields.Integer('Invoiced Amount', compute='_compute_sale_invoiced_amount')
 
     @api.depends('mailing_domain')
     def _compute_sale_quotation_count(self):
-        has_so_access = self.env['sale.order'].check_access_rights('read', raise_exception=False)
-        if not has_so_access:
-            self.sale_quotation_count = 0
-            return
         for mass_mailing in self:
-            mass_mailing.sale_quotation_count = self.env['sale.order'].search_count(mass_mailing._get_sale_utm_domain())
```
Comment (awa-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:23]: Small note: I see that Thibault asked you to change this, maybe it's worth a small note in the commit message?
I had to look for it to see why it changed.

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 23]
```diff
@@ -10,52 +10,59 @@ class MassMailing(models.Model):
...
+        }
 
     def action_redirect_to_invoiced(self):
-        action = self.env["ir.actions.actions"]._for_xml_id("account.action_move_out_invoice_type")
-        moves = self.env['account.move'].search(self._get_sale_utm_domain())
-        action['context'] = {
-            'create': False,
-            'edit': False,
-            'view_no_maturity': True
+        moves = self.env['account.move'].search([('source_id', '=', self.source_id.id)])
```
Comment (awa-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:49]: Why not incorporate this condition in the domain?

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml [hunk 24]
```diff
@@ -22,4 +21,17 @@
     		</xpath>
         </field>
     </record>
-</odoo>
\ No newline at end of file
+
+    <record id="crm_lead_view_search" model="ir.ui.view">
```
Comment (awa-odoo) [inline addons/mass_mailing_crm/views/mailing_mailing_views.xml:25]: Redefining the whole view for this seems a bit overkill.

Can't we simply turn on:
"search_default_month": 1
In the context?
(Yes, for some reason that filter name is "month" ...)

I know it will not group by day as requested but at least it's consistent with the search view.
Same applies for the other inherited views.

Maybe something to re-discuss with @LNAodoo ?

cc @tde-banana-odoo @dbeguin

**Review by awa-odoo — COMMENTED** [review]
Hello @SwagSamaSempai 

Small review as I was passing by.
Added a few minor comments.
I also think we can probably clean all the inherited views? That all seems a bit overkill (see comment in review for details).

Cheers!

File: addons/mass_mailing_crm/views/mailing_mailing_views.xml (hunk 24 above)
Comment (awa-odoo) [reply addons/mass_mailing_crm/views/mailing_mailing_views.xml:25]: Hello again @SwagSamaSempai 

After discussing with Thibault and Luc, we prefer this approach:
https://github.com/odoo/odoo/pull/79166/files#diff-35ba5c308cdb8d8b3dda81053a97d9d12fa686b7fce6241a0db7629b87740b9dR724

Using a bit of context keys.
It does the same thing but reduces the code and avoid extra views.
(Context keys naming should look like: "registration_view_hide_group_by_create_date_week", adapted with view name and granularity of course).

Thanks!

File: addons/mass_mailing_sale/models/mailing_mailing.py (hunk 23 above)
Comment (SwagSamaSempai) [reply addons/mass_mailing_sale/models/mailing_mailing.py:49]: There are two domains because we need to look in two different models. I still changed this part a bit because it wasn't working as intended.

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 25]
```diff
@@ -10,52 +10,71 @@ class MassMailing(models.Model):
...
+    sale_invoiced_amount = fields.Integer('Invoiced Amount', compute='_compute_sale_invoiced_amount')
 
     @api.depends('mailing_domain')
     def _compute_sale_quotation_count(self):
-        has_so_access = self.env['sale.order'].check_access_rights('read', raise_exception=False)
-        if not has_so_access:
-            self.sale_quotation_count = 0
-            return
         for mass_mailing in self:
-            mass_mailing.sale_quotation_count = self.env['sale.order'].search_count(mass_mailing._get_sale_utm_domain())
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:23]: As purpose was to use the source only, can you remove ``_get_sale_utm_domain`` ? I think you removed all usage of it.

**Review by mart-e — CHANGES_REQUESTED** [review]
The variables must be inserted outside of the Markup calls
```py
>>> name = "<R&D>"
>>> Markup("<strong>%s</strong>" % name)
Markup('<strong><R&D></strong>')
>>> Markup("<strong>%s</strong>") % name
Markup('<strong>&lt;R&amp;D&gt;</strong>')
```

File: addons/mass_mailing/models/mailing.py [hunk 26]
```diff
@@ -544,12 +545,19 @@ def _action_view_traces_filtered(self, view_filter):
...
+        helper_header = _("No %s clicked your mailing yet!", recipient)
+        helper_message = _("Link Trackers will measure how many times each link is clicked as well as "
+                           "the proportion of %s who clicked at least once in your mailing.", recipient)
         return {
             'name': model_name,
             'type': 'ir.actions.act_window',
             'view_mode': 'tree',
             'res_model': 'link.tracker',
             'domain': [('mass_mailing_id.id', '=', self.id)],
+            'help': Markup('<p class="o_view_nocontent_smiling_face">%s</p><p>%s</p>' % (
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:558]: Actually the variables should be outside of the Markup call, like ``Markup("<strong>%s</strong>") % "hello"`` (check other Markup use, probably to update accordingly :) )

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 27]
```diff
@@ -10,52 +10,71 @@ class MassMailing(models.Model):
...
+            mass_mailing.sale_quotation_count = self.env['sale.order'].sudo().search_count(
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:19]: Now that sale orders are fetched using a simple domain, can't you do a ``read_group`` instead ? It would allow to have the counters directly with a single query.

File: addons/mass_mailing_sale/models/mailing_mailing.py [hunk 28]
```diff
@@ -10,52 +10,71 @@ class MassMailing(models.Model):
...
-            self.sale_invoiced_amount = 0
-            return
         for mass_mailing in self:
             domain = expression.AND([
-                mass_mailing._get_sale_utm_domain(),
+                [('source_id', '=', mass_mailing.source_id.id)],
                 [('state', 'not in', ['draft', 'cancel'])]
             ])
-            moves = self.env['account.move'].search_read(domain, ['amount_untaxed_signed'])
+            moves = self.env['account.move'].sudo().search_read(domain, ['amount_untaxed_signed'])
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sale/models/mailing_mailing.py:30]: You could probably make a search_read for all mailings (source_id, in, self.source_id.ids). Then in the loop, filter the relevant account moves and sum their untaxed amount. Would allow to have a single search for all mailings to display.

**Review by tde-banana-odoo — COMMENTED** [review]
Review ! Thanks for your PR, some final comments for computed fields. Now that domain is simplified, we can try to batch them in order to speedup their computation :) . Feel free to ping if anything is unclear !

File: addons/mass_mailing/models/mailing.py [hunk 29]
```diff
@@ -566,28 +574,45 @@ def action_view_delivered(self):
...
+        if view_filter == 'reply':
+            found_traces = self.mailing_trace_ids.filtered(lambda trace: trace.trace_status == view_filter)
+            helper_header = _("No %s replied to your mailing yet!", model_name)
+            helper_message = _("To track how many replies this mailing gets, make sure "
+                               "its reply-to address belongs to this database.")
+        elif view_filter == 'bounce':
             found_traces = self.mailing_trace_ids.filtered(lambda trace: trace.trace_status == view_filter)
+            helper_header = _("No %s address bounced yet!", model_name)
+            helper_message = _("Bounce happens when a mailing cannot be delivered (fake address, "
+                               "server issues, ...).<br/>Check each record to see what went wrong.")
```
Comment (mart-e) [inline addons/mass_mailing/models/mailing.py:589]: do you really need this `<br/>` ? It will be escaped with the Markup :thinking:

File: addons/mass_mailing_sale/models/mailing_mailing.py (hunk 27 above)
Comment (SwagSamaSempai) [reply addons/mass_mailing_sale/models/mailing_mailing.py:19]: Like this?
```py
def _compute_sale_quotation_count(self):
    quotation_data = self.env['sale.order'].sudo().read_group(
        [('source_id', 'in', self.source_id.ids)],
        ['source_id'], ['source_id'],
    )
    mapped_data = {quotation['source_id'][0]: quotation['source_id_count'] for quotation in quotation_data}
    for mass_mailing in self:
        mass_mailing.sale_quotation_count = mapped_data.get(mass_mailing.source_id.id, 0)
```

File: addons/mass_mailing_sale/models/mailing_mailing.py (hunk 28 above)
Comment (SwagSamaSempai) [reply addons/mass_mailing_sale/models/mailing_mailing.py:30]: Like this?

```py
def _compute_sale_invoiced_amount(self):
    domain = expression.AND([
        [('source_id', 'in', self.source_id.ids)],
        [('state', 'not in', ['draft', 'cancel'])]
    ])
    moves_data = self.env['account.move'].sudo().read_group(
        domain, ['source_id', 'amount_untaxed_signed'], ['source_id'],
    )
    mapped_data = {datum['source_id'][0]: datum['amount_untaxed_signed'] for datum in moves_data}
    for mass_mailing in self:
        mass_mailing.sale_invoiced_amount = mapped_data.get(mass_mailing.source_id.id, 0)
```

File: addons/mass_mailing/models/mailing.py (hunk 29 above)
Comment (SwagSamaSempai) [reply addons/mass_mailing/models/mailing.py:589]: @LNAodoo Do we?

Comment (LNAodoo) [reply addons/mass_mailing/models/mailing.py]: It would have been great and it looked better but if that is an issue it can be removed :)

File: addons/mass_mailing_sale/models/mailing_mailing.py (hunk 27 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing_sale/models/mailing_mailing.py]: Seems right indeed

File: addons/mass_mailing/models/mailing.py (hunk 29 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing.py]: Otherwise you have to split it in two parts

File: addons/mass_mailing_sale/models/mailing_mailing.py (hunk 28 above)
Comment (SwagSamaSempai) [reply addons/mass_mailing_sale/models/mailing_mailing.py]: It is, I checked :)

File: addons/mass_mailing/models/mailing.py (hunk 29 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing.py]: Indeed. Let us remove the br then.

**Review by mart-e — COMMENTED** [review]
@robodoo override=ci/security
Reason: all html constructed is properly escaped to avoid injection

**Comment by awa-odoo** [comment]
@robodoo r+
